### PR TITLE
Bugfixes May 2024 (Part 1)

### PR DESCRIPTION
**Pull request by qantik**
This PR bundles various bug fixes:

 - #457 (b326b02): Fix correctness errors in `utils.MaxSlice/MinSlice`. Prevent panic in `utils.EqualSlice` 
 - #459 (93154bd): Fix regression in `rlwe.Parameters` marshaller.
 - #466 (34a6325): Update reals tutorial explanation.
 - #458 / #471 (51451c4): Do not mix log/non-log primes in `rlwe.ParametersLiteral` initialisation.
 - #465 (454efab): Use decimal encoding to preserve `rlwe.Scale` to preserve binary size.
 - Bump minimum Go version to 1.21 with the use of `slices`
 
 - (25e88da): New bug reports policy.

### CODE REVIEW

File: core/rlwe/params.go [hunk 1]
```diff
@@ -181,10 +183,11 @@ func NewParametersFromLiteral(paramDef ParametersLiteral) (params Parameters, er
 		paramDef.DefaultScale = s
 	}
 
+	// only allow either (Q or P) or (logQ or logP), do not mix log with non-log
```
Comment (Pro7ech) [inline core/rlwe/params.go:186]: There are use cases where one needs to specify `Q` to ensure that different parameter sets share the same ciphertext modulus (for example switching the ring degree), without having to give specific primes for `P` (i.e. let the library sample them). This prevents such use cases.

File: core/rlwe/scale.go [hunk 2]
```diff
@@ -167,7 +167,7 @@ func (s Scale) Min(s1 Scale) (max Scale) {
 // BinarySize returns the serialized size of the object in bytes.
 // Each value is encoded with .Text('x', ceil(ScalePrecision / log2(10))).
 func (s Scale) BinarySize() int {
-	return 21 + (ScalePrecisionLog10+8)<<1 // 21 for JSON formatting and 2*(8 + ScalePrecisionLog10)
+	return 21 + (ScalePrecisionLog10+6)<<1 // 21 for JSON formatting and 2*(8 + ScalePrecisionLog10)
```
Comment (Pro7ech) [inline core/rlwe/scale.go:170]: Comments have not been updated to reflect the changes

File: core/rlwe/params.go [hunk 3]
```diff
@@ -610,12 +613,15 @@ func (p Parameters) Equal(other *Parameters) (res bool) {
 // MarshalBinary returns a []byte representation of the parameter set.
 // This representation corresponds to the MarshalJSON representation.
 func (p Parameters) MarshalBinary() ([]byte, error) {
```
Comment (Pro7ech) [inline core/rlwe/params.go:615]: Changes and additions to the parameters serialization are still not tested (and this previously caused a merge into main of code not working as expected)

File: core/rlwe/params.go (hunk 1 above)
Comment (qantik) [reply core/rlwe/params.go]: It's fixed.

File: core/rlwe/scale.go (hunk 2 above)
Comment (qantik) [reply core/rlwe/scale.go]: Fixed.

File: core/rlwe/params.go (hunk 3 above)
Comment (qantik) [reply core/rlwe/params.go:641]: Fixed.

**Review by romainbou — APPROVED** [review]

**Pull request merged**